src/lava/utils/weightutils.py

```diff
@@ -17,81 +18,154 @@ class SignMode(Enum):
+        neg_scale = -128 / min_weight if min_weight < 0 else np.inf
         scale = np.min([pos_scale, neg_scale])
     elif sign_mode == SignMode.INHIBITORY:
         scale = -256 / min_weight
     elif sign_mode == SignMode.EXCITATORY:
         scale = 255 / max_weight
 
-    scale_bits = int(np.floor(np.log2(scale)) + is_signed)
+    scale_bits = int(np.floor(np.log2(scale)))
+    if sign_mode == SignMode.MIXED:
+        scale_bits += 1
 
+    weight_exp = -scale_bits
+
+    return weight_exp
+
+
+def _determine_num_weight_bits(weights: np.ndarray,
+                               weight_exp: int,
+                               sign_mode: SignMode) -> int:
     precision_found = False
     n = 8
     while (precision_found is False) and (n > 0):
-        roundingError = np.sum(
-            np.abs(weight / (2**n) - np.round(weight / (2**n)))
+        rounding_error = np.sum(
+            np.abs(weights / (2 ** n) - np.round(weights / (2 ** n)))
         )
-        if roundingError == 0:
+        if rounding_error == 0:
             precision_found = True
         else:
             n -= 1
 
-    n -= is_signed
+    if sign_mode == SignMode.MIXED:
+        n -= 1
 
-    num_weight_bits = 8 - scale_bits - n
-    weight_exponent = -scale_bits
+    num_weight_bits = 8 + weight_exp - n
 
-    weight = np.left_shift(weight.astype(np.int32), int(scale_bits))
+    return num_weight_bits
 
-    if loihi2:
-        weight = weight // (1 << (8 - num_weight_bits))
-        if sign_mode == SignMode.MIXED:
-            weight = weight // 2
-
-    return (
-        weight.astype(int),
-        int(num_weight_bits),
-        int(weight_exponent),
-        sign_mode
-    )
+
+def truncate_weights(weights: np.ndarray,
+                     sign_mode: SignMode,
+                     num_weight_bits: int) -> np.ndarray:
+    """Truncate the least significant bits of the weight matrix given the
+    sign mode and number of weight bits.
+
+    Parameters
+    ----------
+    weights : numpy.ndarray
+        Weight matrix that is to be truncated.
+    sign_mode : SignMode
+        Sign mode to use for truncation. See SignMode class for the
+        correct values.
+    num_weight_bits : int
+        Number of bits to use for the weight matrix.
+
+    Returns
+    -------
+    numpy.ndarray
+        Truncated weight matrix.
+
+    """
+    weights = np.copy(weights).astype(np.int32)
+
+    mixed_flag = 1 if sign_mode == SignMode.MIXED else 0
+    num_truncate_bits = 8 - num_weight_bits + mixed_flag
+
+    # Saturate the weights according to the sign_mode.
+    mixed_flag = int(sign_mode == SignMode.MIXED)
+    excitatory_flag = int(sign_mode == SignMode.EXCITATORY)
+    inhibitory_flag = int(sign_mode == SignMode.INHIBITORY)
+
+    min_wgt = (-2 ** 8) * (mixed_flag + inhibitory_flag)
+    max_wgt = (2 ** 8 - 1) * (mixed_flag + excitatory_flag)
+
+    clipped_weights = np.clip(weights, min_wgt, max_wgt)
```

Comment:
It seems this function not only truncates the least significant bits but also clips the weights. I would prefer to have a separate function for that or at least mention that in the docstring

Reply:
I separated the clipping into its own function and wrote unit tests. See https://github.com/lava-nc/lava/pull/328/commits/7476abe0ebe64dd4fb79c079d553955a2f4f94fd